### PR TITLE
fix(blockfetch): count inverted ranges against stuck-peer recovery

### LINKED ISSUE #3428: blockfetch: count inverted ranges against stuck-peer recovery
`ouroboros/blockfetch.go` rejects an inverted range with `NoBlocks()` but does not call `blockfetchRecordNoBlocksAndMaybeClose`. Oversized ranges and missing start points do call the stuck-peer valve, so a peer can repeat inverted requests without contributing to its no-block response count.

Acceptance criteria:

- Record inverted-range rejections through the same no-block valve as other rejected ranges.
- Close or recycle a peer after the configured repeated-rejection threshold.
- Cover repeated inverted ranges and a valid range control.

### PR DESCRIPTION
**Pull request by cryptodj413**
Fixes #3428

## Problem

`ouroboros/blockfetch.go`'s inverted-range branch (start after end) sent
`NoBlocks` but never called `blockfetchRecordNoBlocksAndMaybeClose`, unlike
the oversized-range and missing-point branches. A peer repeating an inverted
request never counted toward the consecutive-NoBlocks threshold and was
never closed.

## Changes

- Wire the inverted-range branch to the same stuck-peer valve
  (`blockfetchRecordNoBlocksAndMaybeClose`) the other two rejection reasons
  already use, with matching log/close messages.
- Add `TestBlockfetchServerRequestRange_RepeatedInvertedRangeClosesStuckPeer`:
  drives real `MsgRequestRange` traffic over a real `blockfetch.Server`/muxer
  pair and proves the close-eligible WARN fires once the configured
  consecutive-NoBlocks threshold is reached. Confirmed to fail without the fix.
- Add `TestBlockfetchServerRequestRange_RepeatedValidRangeNeverClosesPeer` as
  the control case: a valid, repeated range is never mistaken for an
  inverted-range rejection.

## Checks

- `go build ./...`, `go vet ./...` — clean
- `golangci-lint run ./ouroboros/...`, `nilaway ./ouroboros/...`,
  `modernize ./ouroboros/...` — clean
- `make import-boundaries`, `make golines`, `make docs-parity` — clean, no diff
- `make govulncheck` — no vulnerabilities in reachable code
- `go test -race ./ouroboros/...` — pass
- `make test` (full module, race-enabled, all 84 packages including
  `internal/test/conformance`) — pass, no failures
- `ARCHITECTURE.md`/`DATABASE.md` checked: unaffected. This wires an
  already-documented mechanism (repeated NoBlocks closes the stuck peer,
  documented in ARCHITECTURE.md) into a third branch that already existed;
  it does not change component responsibilities, EventBus topics, or
  lifecycle/concurrency behavior.

## Skipped

- `internal/test/devnet/run-tests.sh` — Docker unavailable in the validation
  environment (permission denied on `docker.sock`). The change is confined to
  the blockfetch server's request-validation rejection path (an edge case a
  healthy sync never exercises), not block serving/streaming/consensus.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Fixes #3428 by counting inverted block-range requests toward stuck-peer recovery. Previously, requests with a start point after the end point returned `NoBlocks` without updating the consecutive counter; they now use the existing recovery path, so peers that repeat them can be closed after the configured threshold.

- Adds real blockfetch server/muxer coverage for the threshold behavior and a valid-range control case.
- Shares the `net.Pipe`/muxer server-peer test harness between blockfetch and Leios fetch tests.

<sup>Written for commit 46f82042e742f2bd9cb5b2497506c9731137ccb1. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/dingo/pull/3812?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Bug Fixes**
  - Invalid inverted block-range requests are now tracked consistently with other repeated invalid requests.
  - Peers that repeatedly send invalid inverted ranges can now be closed after reaching the configured threshold.
  - Valid block-range requests continue to be processed normally without triggering peer closure.

- **Tests**
  - Added coverage for repeated invalid ranges closing a peer.
  - Added coverage confirming valid ranges do not cause peer closure.

### CODE REVIEW

File: ouroboros/blockfetch_test.go [hunk 1]
```diff
@@ -718,6 +752,188 @@ func TestBlockfetchRecordNoBlocks_CleanupResetsCounter(t *testing.T) {
...
+				strings.Contains(logBuf.String(), closeWarnMsg),
+				"request %d should not yet reach the close threshold",
+				i,
+			)
+		} else {
+			// At threshold, the close-eligible WARN is logged after NoBlocks()
+			// is enqueued, on the server's own goroutine, with no ordering
+			// guarantee relative to the wire bytes readResponse observed --
+			// poll instead of asserting immediately.
+			testutil.WaitForCondition(
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ouroboros/blockfetch_test.go:836]: Valid — fixed in 50a20e1. Renamed to `TestBlockfetchServerRequestRange_RepeatedInvertedRangeReachesCloseThreshold` and reworded the doc comment: this test isolates whether the inverted-range branch now feeds the shared no-blocks valve and reaches its close-eligible threshold. It does not assert an actual `conn.Close()`, since `connManager` is nil here and `blockfetchRecordNoBlocksAndMaybeClose` only calls `closeBlockfetchConnection` when `connManager` is non-nil. The generic close call is already covered by `TestBlockfetchServerSendBatch_ClosesConnectionWhenSendDrainStalls` and `TestReportBlockfetchServerAsyncError_ClosesConnection`.

File: ouroboros/blockfetch_test.go [hunk 2]
```diff
@@ -718,6 +752,188 @@ func TestBlockfetchRecordNoBlocks_CleanupResetsCounter(t *testing.T) {
...
+		server.Protocol.Stop()
+		m.Stop()
+		_ = serverConn.Close()
+		_ = peerConn.Close()
+	})
+	return &blockfetchServerPeer{peerConn: peerConn, errChan: errChan}
+}
+
+// send writes msg to the server as a block-fetch request segment.
+func (p *blockfetchServerPeer) send(t *testing.T, msg protocol.Message) {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ouroboros/blockfetch_test.go]: Valid observation, but not applying it in this PR. This package already carries multiple un-deduplicated per-protocol test-peer harnesses of this same net.Pipe+muxer shape (`chainsyncServerFixture` in chainsync_server_fixture_test.go, `leiosFetchServerPeer` in leios_range_responder_test.go, now `blockfetchServerPeer`) — each protocol's harness owns its own `NewConfig`/`NewServer` types and message helpers, which don't share a common interface to generalize over without adding indirection. A shared low-level net.Pipe/muxer/send/readResponse helper is a reasonable follow-up across all three, but it's a repo-wide test-infra refactor, not something scoped to this bug fix.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
Reviewed current head 50a20e1. Cubic’s P3 finding is valid: blockfetch_test.go adds a net.Pipe/muxer peer harness duplicating leios_range_responder_test.go. Factor the shared harness before approval; the behavioral fix and CI otherwise look sound.

**Comment by cryptodj413** [comment]
Addressed in 46f8204: factored the net.Pipe/muxer server-peer harness (blockfetchServerPeer duplicated leiosFetchServerPeer) into a shared `muxerServerPeer` type in `ouroboros/muxer_server_peer_test.go`. Each protocol still builds its own Config/Server (blockfetch.NewConfig/NewServer vs leiosfetch's, which have different signatures), but the net.Pipe/muxer plumbing and send/readResponse wire mechanics are now shared and used by both `blockfetch_test.go` and `leios_range_responder_test.go`.

Validated: `go build ./...`, `go vet ./...`, golangci-lint/nilaway/modernize, `make import-boundaries`/`golines`/`docs-parity`, `go test -race ./ouroboros/...`, and full `make test` (84/84 packages) — all clean. CI green on 46f8204.

**Review by wolf31o2 — APPROVED** [review]
